torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -18,74 +17,55 @@ namespace {
+
  public:
-  // Write-After-Read race conditions are only found within for-loops.
-  // Sync nodes are inserted directly into the for-loops.
-  // The expressions are modified in-place and exprs is const.
-  static void InsertSyncs(const std::vector<Expr*>& exprs) {
+  //! Write-After-Read race conditions are only found within for-loops.
+  //! Sync nodes are inserted directly into the for-loops.
+  //! The expressions are modified in-place and exprs is const.
+  static void insertSyncs(const std::vector<kir::Expr*>& exprs) {
     LocalSyncInserter sync_inserter;
     for (auto expr : exprs) {
       sync_inserter.handle(expr);
     }
   }
 
-  const std::unordered_set<const TensorView*>& initial() const {
+  const auto& initial() const {
     return initial_;
   }
 
-  const std::unordered_set<const TensorView*>& final() const {
+  const auto& final() const {
     return final_;
   }
 
-  const std::unordered_set<const TensorView*>& all_smem_inputs() const {
+  const auto& all_smem_inputs() const {
     return all_smem_inputs_;
   }
 
-  const std::unordered_set<const TensorView*>& all_smem_outputs() const {
+  const auto& all_smem_outputs() const {
     return all_smem_outputs_;
   }
 
-  const std::unordered_set<unsigned int>& all_aliased_allocations() const {
-    return all_alias_allocations_;
-  }
-
  private:
-  explicit LocalSyncInserter(
-      const std::unordered_set<unsigned int>* parent_alias_allocations =
-          nullptr) {
-    if (parent_alias_allocations != nullptr) {
-      all_alias_allocations_.insert(
-          parent_alias_allocations->begin(), parent_alias_allocations->end());
-    }
-  }
-
-  void handle(Expr* expr) final {
+  // TODO(kir): this is a place where a mutable IR visitor may be appropriate
+  void handle(kir::Expr* expr) {
     if (ir_utils::isTVOp(expr)) {
-      // Alias Allocations
-      const auto output = expr->output(0)->as<TensorView>();
-
       // For this SyncInserter
-      (!initial_sync_) ? hasOutputSmemExpr(expr, initial_)
-                       : hasInputSmemExpr(expr, final_);
+      initial_sync_ ? addInputSmemTvs(expr, final_)
+                    : addOutputSmemTvs(expr, initial_);
 
       // For parent SyncInserter
-      hasOutputSmemExpr(expr, all_smem_outputs_);
-      hasInputSmemExpr(expr, all_smem_inputs_);
-    } else {
-      OptOutDispatch::handle(expr);
-    }
-  }
-
-  void handle(kir::Allocate* a) final {
-    if (a->buffer()->getValType().value() == ValType::KirTensorView &&
-        a->alias() != nullptr && a->getMemoryType() == MemoryType::Shared) {
-      auto tv = a->buffer()->as<kir::TensorView>()->fuserTv();
-      all_alias_allocations_.insert(tv->name());
+      addOutputSmemTvs(expr, all_smem_outputs_);
+      addInputSmemTvs(expr, all_smem_inputs_);
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      handle(ite);
+    } else if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      handle(for_loop);
```

Comment:
Do we need explicit dispatching?

Reply:
Same situation as the comment in lower_alias_memory.cpp. There's a TODO comment on top of this method to revisit the dispatch, but right now it's required.